server/src/main/java/org/elasticsearch/index/engine/CombinedDocValues.java

```diff
@@ -47,6 +47,7 @@
     long docVersion(int segmentDocId) throws IOException {
         assert versionDV.docID() < segmentDocId;
         if (versionDV.advanceExact(segmentDocId) == false) {
+            assert false : "DocValues for field [" + VersionFieldMapper.NAME + "] is not found";
```

Comment:
what is the point of doubling the exception with an assertion?

Reply:
We prefer to bring a node down in tests but not in production when this invariant does not hold.  We use this approach in other places, for example, https://github.com/elastic/elasticsearch/blob/6e2f7b4b0842654ea9db8837a935947a6da49aa4/server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java#L276-L277.

Reply:
Thanks for explaining, it makes sense.